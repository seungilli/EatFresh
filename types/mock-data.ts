import { HistoryType, Meal } from "./data";

export const mockMeal1: Meal = {
    idMeal: "67890",
    strMeal: "Delicious Pasta",
    strMealThumb: "https://via.placeholder.com/150",
    strCategory: "Pasta",
    strArea: "Italian",
    strDrinkAlternate: null,
    strInstructions: "Boil pasta. Prepare sauce. Mix and serve.",
    strTags: "Quick,Easy",
    strYoutube: "https://youtube.com/test-video-pasta",
    strIngredient1: "Pasta",
    strIngredient2: "Tomato Sauce",
    strIngredient3: "Cheese",
    strIngredient4: "Garlic",
    strIngredient5: "Basil",
    strIngredient6: "Salt",
    strIngredient7: "Pepper",
    strIngredient8: "Olive Oil",
    strIngredient9: "Onion",
    strIngredient10: null,
    strIngredient11: null,
    strIngredient12: null,
    strIngredient13: null,
    strIngredient14: null,
    strIngredient15: null,
    strIngredient16: null,
    strIngredient17: null,
    strIngredient18: null,
    strIngredient19: null,
    strIngredient20: null,
    strMeasure1: "200 grams",
    strMeasure2: "1 cup",
    strMeasure3: "100 grams",
    strMeasure4: "2 cloves",
    strMeasure5: "5 leaves",
    strMeasure6: "to taste",
    strMeasure7: "to taste",
    strMeasure8: "2 tbsp",
    strMeasure9: "1 medium",
    strMeasure10: null,
    strMeasure11: null,
    strMeasure12: null,
    strMeasure13: null,
    strMeasure14: null,
    strMeasure15: null,
    strMeasure16: null,
    strMeasure17: null,
    strMeasure18: null,
    strMeasure19: null,
    strMeasure20: null,
    strSource: "Family Recipe",
    strImageSource: "https://via.placeholder.com/image-source1",
    strCreativeCommonsConfirmed: "Yes",
    dateModified: "2023-05-15",
};


export const mockMeal2: Meal = {
    idMeal: "54321",
    strMeal: "Spicy Chicken Curry",
    strMealThumb: "https://via.placeholder.com/200",
    strCategory: "Curry",
    strArea: "Indian",
    strDrinkAlternate: "Lassi",
    strInstructions: "Marinate chicken. Cook with spices. Serve hot.",
    strTags: "Spicy,Dinner",
    strYoutube: "https://youtube.com/test-video-curry",
    strIngredient1: "Chicken",
    strIngredient2: "Curry Powder",
    strIngredient3: "Yogurt",
    strIngredient4: "Garlic",
    strIngredient5: "Ginger",
    strIngredient6: "Onion",
    strIngredient7: "Tomato",
    strIngredient8: "Coriander",
    strIngredient9: "Chili Powder",
    strIngredient10: "Salt",
    strIngredient11: "Pepper",
    strIngredient12: "Oil",
    strIngredient13: "Cumin Seeds",
    strIngredient14: "Turmeric",
    strIngredient15: "Cardamom",
    strIngredient16: "Cinnamon Stick",
    strIngredient17: null,
    strIngredient18: null,
    strIngredient19: null,
    strIngredient20: null,
    strMeasure1: "500 grams",
    strMeasure2: "2 tbsp",
    strMeasure3: "1 cup",
    strMeasure4: "3 cloves",
    strMeasure5: "1 inch",
    strMeasure6: "2 medium",
    strMeasure7: "2 large",
    strMeasure8: "to garnish",
    strMeasure9: "1 tsp",
    strMeasure10: "to taste",
    strMeasure11: "to taste",
    strMeasure12: "3 tbsp",
    strMeasure13: "1 tsp",
    strMeasure14: "1/2 tsp",
    strMeasure15: "2 pods",
    strMeasure16: "1 stick",
    strMeasure17: null,
    strMeasure18: null,
    strMeasure19: null,
    strMeasure20: null,
    strSource: "Traditional Indian Cookbook",
    strImageSource: "https://via.placeholder.com/image-source2",
    strCreativeCommonsConfirmed: "No",
    dateModified: "2022-12-01",
};

export const mockHistoryItem1: HistoryType = {
    time: "2023-11-23T12:00:00Z",
    searchedText: "History Item 1",
};

export const mockHistoryItem2: HistoryType = {
    time: "2023-11-24T14:00:00Z",
    searchedText: "History Item 2",
};


