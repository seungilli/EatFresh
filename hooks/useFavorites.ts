import { useState, useEffect, useCallback } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Meal } from '../types/data';

const useFavorites = () => {
    const [favorites, setFavorites] = useState<Meal[]>([]);

    const fetchFavorites = useCallback(async () => {
        try {
            const value = await AsyncStorage.getItem('favorite');
            const parsedFavorites = value != null ? JSON.parse(value) : [];
            setFavorites(Array.isArray(parsedFavorites) ? parsedFavorites : []);
        } catch (error) {
            console.error('Error fetching favorites:', error);
        }
    }, []);

    const addFavorite = useCallback(
        async (item: Meal) => {
            try {
                const existingFavorites = await AsyncStorage.getItem('favorite');
                let updatedFavorites = existingFavorites ? JSON.parse(existingFavorites) : [];
                if (!Array.isArray(updatedFavorites)) {
                    updatedFavorites = [];
                }
                updatedFavorites.push(item);
                await AsyncStorage.setItem('favorite', JSON.stringify(updatedFavorites));
                await fetchFavorites();
            } catch (error) {
                console.error('Error adding favorite:', error);
            }
        },
        [fetchFavorites]
    );

    const removeFavorite = useCallback(
        async (id: string) => {
            try {
                const existingFavorites = await AsyncStorage.getItem('favorite');
                let updatedFavorites = existingFavorites ? JSON.parse(existingFavorites) : [];
                if (Array.isArray(updatedFavorites)) {
                    updatedFavorites = updatedFavorites.filter((meal: Meal) => meal.idMeal !== id);
                    await AsyncStorage.setItem('favorite', JSON.stringify(updatedFavorites));
                    await fetchFavorites();
                }
            } catch (error) {
                console.error('Error removing favorite:', error);
            }
        },
        [fetchFavorites]
    );

    const clearFavorites = useCallback(async () => {
        try {
            await AsyncStorage.removeItem('favorite');
            await fetchFavorites();
        } catch (error) {
            console.error('Error clearing favorites:', error);
        }
    }, [fetchFavorites]);

    useEffect(() => {
        fetchFavorites();
    }, [fetchFavorites]);

    return {
        favorites,
        addFavorite,
        removeFavorite,
        clearFavorites,
        refreshFavorites: fetchFavorites,
    };
};

export default useFavorites;
