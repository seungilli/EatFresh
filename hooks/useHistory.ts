import { useState, useEffect, useCallback } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { HistoryType } from '@/types/data';

const useHistory = () => {
    const [history, setHistory] = useState<HistoryType[]>([]);

    const fetchHistory = useCallback(async () => {
        try {
            const value = await AsyncStorage.getItem('history');
            const parsedHistory = value != null ? JSON.parse(value) : [];
            setHistory(Array.isArray(parsedHistory) ? parsedHistory : []);
        } catch (error) {
            console.error('Error fetching history:', error);
        }
    }, []);

    const addHistoryItem = useCallback(
        async (item: HistoryType) => {
            try {
                const existingHistory = await AsyncStorage.getItem('history');
                let updatedHistory = existingHistory ? JSON.parse(existingHistory) : [];
                if (!Array.isArray(updatedHistory)) {
                    updatedHistory = [];
                }
                updatedHistory.push(item);
                await AsyncStorage.setItem('history', JSON.stringify(updatedHistory));
                await fetchHistory();
            } catch (error) {
                console.error('Error adding history:', error);
            }
        },
        [fetchHistory]
    );

    const clearHistory = useCallback(async () => {
        try {
            await AsyncStorage.removeItem('history');
            await fetchHistory();
        } catch (error) {
            console.error('Error clearing history:', error);
        }
    }, [fetchHistory]);

    useEffect(() => {
        fetchHistory();
    }, [fetchHistory]);

    return {
        history,
        addHistoryItem,
        clearHistory,
        refreshHistory: fetchHistory,
    };
};

export default useHistory;
