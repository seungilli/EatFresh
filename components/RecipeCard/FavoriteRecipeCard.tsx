import { Meal, RecipeCardProps } from "@/hooks/useFetchData";
import { Ionicons } from "@expo/vector-icons";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { Link } from "expo-router";
import React from "react";
import { StyleSheet, Text, View, Image, TouchableOpacity } from "react-native";
import { Card } from "react-native-paper";

const styles = StyleSheet.create({
  card: {
    marginVertical: 8,
    padding: 8,
    borderRadius: 8,
    elevation: 2,
  },
  container: {
    flexDirection: "row",
    alignItems: "center",
  },
  image: {
    width: 100,
    height: 100,
    borderRadius: 8,
  },
  contentContainer: {
    flex: 1,
    marginHorizontal: 12,
  },
  title: {
    fontSize: 16,
    fontWeight: "bold",
    marginBottom: 4,
  },
  tagsContainer: {
    flexDirection: "row",
    gap: 8,
  },
  tag: {
    paddingHorizontal: 8,
    paddingVertical: 4,
    borderRadius: 16,
    fontSize: 12,
    fontWeight: "bold",
  },
  heartIconContainer: {
    padding: 4,
  },
  heartIcon: {
    fontSize: 18,
  },
});

type FavoriteRecipeCardProps = {
  item: Meal;
};

export default function FavoriteRecipeCard({ item }: FavoriteRecipeCardProps) {
  const setItem = async (item: Meal) => {
    try {
      const existingFavorites = await AsyncStorage.getItem("favorite");
      let favorites = existingFavorites ? JSON.parse(existingFavorites) : [];
      if (!Array.isArray(favorites)) {
        favorites = [];
      }
      favorites.push(item);
      await AsyncStorage.setItem("favorite", JSON.stringify(favorites));
    } catch (error) {
      console.error("Error setting item:", error);
    }
  };

  return (
    <Card style={styles.card}>
      <Link href={`/details?recipeId=${item.idMeal}`}>
        <View style={styles.container}>
          {item.strMealThumb ? (
            <Image source={{ uri: item.strMealThumb }} style={styles.image} />
          ) : null}
          <View style={styles.contentContainer}>
            <Text style={styles.title}>{item.strMeal}</Text>
            <View style={styles.tagsContainer}>
              <Text style={styles.tag}>{item.strCategory}</Text>
              <Text style={styles.tag}>{item.strArea}</Text>
            </View>
          </View>
          <TouchableOpacity
            onPress={() => setItem(item)}
            style={styles.heartIconContainer}
          >
            <Ionicons name="heart-outline" size={20} color="grey" />
          </TouchableOpacity>
        </View>
      </Link>
    </Card>
  );
}
