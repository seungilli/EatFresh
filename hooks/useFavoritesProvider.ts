import React, { createContext, useContext, ReactNode } from "react";
import useFavorites from "./useFavorites";

type FavoritesContextType = ReturnType<typeof useFavorites>;

const FavoritesContext = createContext<FavoritesContextType | undefined>(
    undefined
);

export const FavoritesProvider = ({ children }: { children: ReactNode }) => {
    const favorites = useFavorites();

    return React.createElement(
        FavoritesContext.Provider,
        { value: favorites },
        children
    );
};

export const useFavoritesContext = (): FavoritesContextType => {
    const context = useContext(FavoritesContext);
    if (!context) {
        throw new Error(
            "useFavoritesContext must be used within a FavoritesProvider"
        );
    }
    return context;
};
