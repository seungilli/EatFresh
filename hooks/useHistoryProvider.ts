import React, { createContext, useContext, ReactNode } from "react";
import useHistory from "./useHistory";

type HistoryContextType = ReturnType<typeof useHistory>;

const HistoryContext = createContext<HistoryContextType | undefined>(
    undefined
);

export const HistoryProvider = ({ children }: { children: ReactNode }) => {
    const history = useHistory();

    return React.createElement(
        HistoryContext.Provider,
        { value: history },
        children
    );
};

export const useHistoryContext = (): HistoryContextType => {
    const context = useContext(HistoryContext);
    if (!context) {
        throw new Error(
            "useHistoryContext must be used within a HistoryProvider"
        );
    }
    return context;
};
