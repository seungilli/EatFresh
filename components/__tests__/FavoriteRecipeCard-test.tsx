import React from "react";
import renderer, { act } from "react-test-renderer";
import { TouchableOpacity, Text, Image } from "react-native";
import { Link } from "expo-router";
import { Ionicons } from "@expo/vector-icons";
import FavoriteRecipeCard from "../RecipeCard/FavoriteRecipeCard";
import { Meal } from "@/types/data";

jest.mock("@expo/vector-icons", () => ({
  Ionicons: "Ionicons",
}));

jest.mock("@/hooks/useFavoritesProvider", () => ({
  useFavoritesContext: jest.fn(),
}));

describe("<FavoriteRecipeCard />", () => {
  const mockFavoritesContext =
    require("@/hooks/useFavoritesProvider").useFavoritesContext;

  const mockMeal: Meal = {
    idMeal: "12345",
    strMeal: "Test Meal",
    strMealThumb: "https://via.placeholder.com/100",
    strCategory: "Test Category",
    strArea: "Test Area",
    strDrinkAlternate: null,
    strInstructions: "Test instructions",
    strTags: "Test tags",
    strYoutube: "Test Video",
    strIngredient1: "Ingredient1",
    strIngredient2: "Ingredient2",
    strIngredient3: "Ingredient3",
    strIngredient4: "Ingredient4",
    strIngredient5: "Ingredient5",
    strIngredient6: "Ingredient6",
    strIngredient7: "Ingredient7",
    strIngredient8: "Ingredient8",
    strIngredient9: "Ingredient9",
    strIngredient10: "Ingredient10",
    strIngredient11: "Ingredient11",
    strIngredient12: "Ingredient12",
    strIngredient13: "Ingredient13",
    strIngredient14: "Ingredient14",
    strIngredient15: "Ingredient15",
    strIngredient16: "Ingredient16",
    strIngredient17: "Ingredient17",
    strIngredient18: "Ingredient18",
    strIngredient19: "Ingredient19",
    strIngredient20: "Ingredient20",
    strMeasure1: "1 cup",
    strMeasure2: "2 tbsp",
    strMeasure3: "3 tsp",
    strMeasure4: "4 oz",
    strMeasure5: "5 grams",
    strMeasure6: "6 ml",
    strMeasure7: "7 slices",
    strMeasure8: "8 pieces",
    strMeasure9: "9 cups",
    strMeasure10: "10 tbsp",
    strMeasure11: "11 tsp",
    strMeasure12: "12 oz",
    strMeasure13: "13 grams",
    strMeasure14: "14 ml",
    strMeasure15: "15 slices",
    strMeasure16: "16 pieces",
    strMeasure17: "17 cups",
    strMeasure18: "18 tbsp",
    strMeasure19: "19 tsp",
    strMeasure20: "20 oz",
    strSource: "Test Source",
    strImageSource: "Test Image Source",
    strCreativeCommonsConfirmed: "Yes",
    dateModified: "2023-01-01",
  };

  const mockAddFavorite = jest.fn();
  const mockRemoveFavorite = jest.fn();

  beforeEach(() => {
    mockFavoritesContext.mockReturnValue({
      favorites: [],
      addFavorite: mockAddFavorite,
      removeFavorite: mockRemoveFavorite,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render", () => {
    const tree = renderer.create(<FavoriteRecipeCard item={mockMeal} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should render the meal image and title", () => {
    const component = renderer.create(<FavoriteRecipeCard item={mockMeal} />);
    const root = component.root;

    const image = root.findByType(Image);
    expect(image.props.source.uri).toBe(mockMeal.strMealThumb);

    const textNodes = root.findAllByType(Text);
    const titleNode = textNodes.find(
      (node) => node.props.children === mockMeal.strMeal
    );

    expect(titleNode).toBeDefined();
    expect(titleNode?.props.children).toBe(mockMeal.strMeal);
  });

  it("should render category and area tags", () => {
    const component = renderer.create(<FavoriteRecipeCard item={mockMeal} />);
    const root = component.root;

    const textNodes = root.findAllByType(Text);

    const categoryNode = textNodes.find(
      (node) => node.props.children === mockMeal.strCategory
    );
    const areaNode = textNodes.find(
      (node) => node.props.children === mockMeal.strArea
    );

    expect(categoryNode).toBeDefined();
    expect(categoryNode?.props.children).toBe(mockMeal.strCategory);

    expect(areaNode).toBeDefined();
    expect(areaNode?.props.children).toBe(mockMeal.strArea);
  });

  it("should navigate to the details page when clicked", () => {
    const component = renderer.create(<FavoriteRecipeCard item={mockMeal} />);
    const root = component.root;

    const link = root.findByType(Link);
    expect(link.props.href).toBe(`/details?recipeId=${mockMeal.idMeal}`);
  });

  it("should display an outline heart icon when not in favorites", () => {
    const component = renderer.create(<FavoriteRecipeCard item={mockMeal} />);
    const root = component.root;

    const heartIcon = root.findByType(Ionicons);
    expect(heartIcon.props.name).toBe("heart-outline");
  });

  it("should display a filled heart icon when in favorites", () => {
    mockFavoritesContext.mockReturnValue({
      favorites: [mockMeal],
      addFavorite: mockAddFavorite,
      removeFavorite: mockRemoveFavorite,
    });

    const component = renderer.create(<FavoriteRecipeCard item={mockMeal} />);
    const root = component.root;

    const heartIcon = root.findByType(Ionicons);
    expect(heartIcon.props.name).toBe("heart");
  });

  it("should call addFavorite when heart icon is pressed", () => {
    const component = renderer.create(<FavoriteRecipeCard item={mockMeal} />);
    const root = component.root;

    const heartButton = root.findByType(TouchableOpacity);

    act(() => {
      heartButton.props.onPress();
    });

    expect(mockAddFavorite).toHaveBeenCalledWith(mockMeal);
    expect(mockRemoveFavorite).not.toHaveBeenCalled();
  });

  it("should call removeFavorite when heart icon is pressed", () => {
    mockFavoritesContext.mockReturnValue({
      favorites: [mockMeal],
      addFavorite: mockAddFavorite,
      removeFavorite: mockRemoveFavorite,
    });

    const component = renderer.create(<FavoriteRecipeCard item={mockMeal} />);
    const root = component.root;

    const heartButton = root.findByType(TouchableOpacity);

    act(() => {
      heartButton.props.onPress();
    });

    expect(mockRemoveFavorite).toHaveBeenCalledWith(mockMeal.idMeal);
    expect(mockAddFavorite).not.toHaveBeenCalled();
  });
});
