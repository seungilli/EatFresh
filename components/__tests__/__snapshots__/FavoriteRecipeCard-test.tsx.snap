// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteRecipeCard Component renders correctly with meal data 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgba(247, 243, 249, 1)",
      "borderRadius": 8,
      "marginVertical": 8,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="card-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(247, 243, 249, 1)",
        "borderRadius": 8,
        "elevation": 2,
        "flex": undefined,
        "padding": 8,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="card-container"
  >
    <View
      style={
        [
          {
            "flexShrink": 1,
          },
          undefined,
        ]
      }
      testID="card"
    >
      <Text
        borderRadiusStyles={
          {
            "borderRadius": 8,
          }
        }
        href="/details?recipeId=12345"
        index={0}
        onPress={[Function]}
        role="link"
        siblings={[]}
        total={1}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://via.placeholder.com/100",
              }
            }
            style={
              {
                "borderRadius": 8,
                "height": 100,
                "width": 100,
              }
            }
          />
          <View
            style={
              {
                "flex": 1,
                "marginHorizontal": 12,
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginBottom": 4,
                }
              }
            >
              Test Meal
            </Text>
            <View
              style={
                {
                  "flexDirection": "row",
                  "gap": 8,
                }
              }
            >
              <Text
                style={
                  {
                    "backgroundColor": "#cbd38f",
                    "borderRadius": 16,
                    "fontSize": 12,
                    "fontWeight": "bold",
                    "paddingHorizontal": 8,
                    "paddingVertical": 4,
                  }
                }
              >
                Test Category
              </Text>
              <Text
                style={
                  {
                    "backgroundColor": "#cbd38f",
                    "borderRadius": 16,
                    "fontSize": 12,
                    "fontWeight": "bold",
                    "paddingHorizontal": 8,
                    "paddingVertical": 4,
                  }
                }
              >
                Test Area
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
                "padding": 4,
              }
            }
          >
            <Ionicons
              color="#d2d2d2"
              name="heart-outline"
              size={30}
            />
          </View>
        </View>
      </Text>
    </View>
  </View>
</View>
`;
